src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -316,29 +333,33 @@ private static ImmutableArray<TextChangeRange> Merge(ImmutableArray<TextChangeRa
                             AddRange(list, oldChange);
                             oldDelta = oldDelta - oldChange.Span.Length + oldChange.NewLength;
                             oldIndex++;
+                            needNextOldChange = true;
                             goto nextOldChange;
                         }
                         else if (newChange.Span.Length <= oldChange.NewLength)
                         {
                             // new change deletes fewer characters than old change inserted
-                            // add new change insertion, then the remaining trailing characters of the old change insertion
-                            AddRange(list, new TextChangeRange(oldChange.Span, oldChange.NewLength + newChange.NewLength - newChange.Span.Length));
-                            oldDelta = oldDelta - oldChange.Span.Length + oldChange.NewLength - newChange.Span.Length;
-                            oldIndex++;
+                            // apply the new change, and adjust the old change
+                            var adjustedNewChange = new TextChangeRange(new TextSpan(newChange.Span.Start - oldDelta, newChange.Span.Length), newChange.NewLength);
+                            AddRange(list, adjustedNewChange);
+                            oldChange = new TextChangeRange(new TextSpan(oldChange.Span.Start + adjustedNewChange.Span.Length, oldChange.Span.Length - adjustedNewChange.Span.Length), oldChange.NewLength - newChange.Span.Length);
                             newIndex++;
+                            needNextNewChange = true;
                             goto nextNewChange;
                         }
                         else
                         {
-                            // delete as much from old change as new change can
-                            // a new change deletion is a reduction in the old change insertion
-                            var oldChangeReduction = Math.Min(oldChange.NewLength, newChange.Span.Length);
-                            AddRange(list, new TextChangeRange(oldChange.Span, oldChange.NewLength - oldChangeReduction));
-                            oldDelta = oldDelta - oldChange.Span.Length + (oldChange.NewLength - oldChangeReduction);
```

Comment:
📝 Due to the `if` statement conditions, this was equivalent to:

```csharp
var oldChangeReduction = oldChange.NewLength;
AddRange(list, new TextChangeRange(oldChange.Span, 0));
oldDelta = oldDelta - oldChange.Span.Length;
```

📝 I don't think this delta calculation was correct